src/fluree/db/json_ld/transact.cljc

```diff
@@ -77,37 +77,41 @@
 
 (defn add-property
   "Adds property. Parameters"
-  [sid pid {shacl-dt :dt, validate-fn :validate-fn} check-retracts? list? {:keys [value] :as v-map}
-   {:keys [t db-before] :as tx-state}]
+  [sid pid shacl-dt shape->p-shapes check-retracts? list? {:keys [value] :as v-map}
+   {:keys [t db-before subj-mods] :as tx-state}]
   (go-try
     (let [retractions (when check-retracts? ;; don't need to check if generated pid during this transaction
                         (->> (<? (query-range/index-range db-before :spot = [sid pid]))
                              (map #(flake/flip-flake % t))))
           m           (when list?
                         {:i (-> v-map :idx last)})
+
           flakes      (cond
                         ;; a new node's data is contained, process as another node then link to this one
                         (jld-reify/node? v-map)
-                        (do
-                          (when validate-fn
-                            (or (validate-fn v-map)
-                                (throw (ex-info (str "Node did not pass SHACL validation: " v-map)
-                                                {:status 400, :error :db/shacl-validation}))))
-                          (let [[node-sid node-flakes] (<? (json-ld-node->flakes v-map tx-state pid))]
-                            (conj node-flakes (flake/create sid pid node-sid const/$xsd:anyURI t true m))))
+                        (let [[node-sid node-flakes] (<? (json-ld-node->flakes v-map tx-state pid))]
+                          (conj node-flakes (flake/create sid pid node-sid const/$xsd:anyURI t true m)))
 
                         ;; a literal value
                         (and (some? value) (not= shacl-dt const/$xsd:anyURI))
                         (let [[value* dt] (datatype/from-expanded v-map shacl-dt)]
-                          (when validate-fn
-                            (or (validate-fn value*)
-                                (throw (ex-info (str "Value did not pass SHACL validation: " value)
-                                                {:status 400 :error :db/shacl-validation}))))
                           [(flake/create sid pid value* dt t true m)])
 
                         :else
                         (throw (ex-info (str "JSON-LD value must be a node or a value, instead found ambiguous value: " v-map)
-                                        {:status 400 :error :db/invalid-transaction})))]
+                                        {:status 400 :error :db/invalid-transaction})))
+          [valid? err-msg] (shacl/coalesce-validation-results
+                             (mapcat (fn [[shape-id p-shapes]]
+                                       ;; register the validated pid so we can enforce the sh:closed constraint later
+                                       (swap! subj-mods update-in [:shape->validated-properties shape-id]
```

Comment:
Using `mapcat` with side effects gives me some pause. I think it works ok here because we're passing it to a fn that calls `reduce`, but in principle it feels a little iffy to me. The more reliably eager alternative would be to use `into` with a `mapcat` transducer